README.md
# NAME

HTTP::Tiny::Paranoid - A safer HTTP::Tiny

# SYNOPSIS

    # use just like HTTP::Tiny
    use HTTP::Tiny::Paranoid;
    my $response = HTTP::Tiny::Paranoid->new->get('http://example.com');

    # block or whitelist specific hosts
    # delegates to Net::DNS::Paranoid
    HTTP::Tiny::Paranoid->blocked_hosts([...]);
    HTTP::Tiny::Paranoid->whitelisted_hosts([...]);

# DESCRIPTION

This module is a subclass of HTTP::Tiny that performs exactly one additional
function: before connecting, it passes the hostname to
[Net::DNS::Paranoid](https://metacpan.org/pod/Net::DNS::Paranoid). If the hostname is rejected, then the request
is aborted before a connect is even attempted.

By default, [Net::DNS::Paranoid](https://metacpan.org/pod/Net::DNS::Paranoid) rejects connections to private network
ranges. The blocklist & whitelist can be manipulated using the `blocked_hosts`
and `whitelisted_hosts` class methods.

# SEE ALSO

- [HTTP::Tiny](https://metacpan.org/pod/HTTP::Tiny)
- [Net::DNS::Paranoid](https://metacpan.org/pod/Net::DNS::Paranoid)
- [LWPx::ParanoidAgent](https://metacpan.org/pod/LWPx::ParanoidAgent)
- [LWP::UserAgent::Paranoid](https://metacpan.org/pod/LWP::UserAgent::Paranoid)

# SUPPORT

## Bugs / Feature Requests

Please report any bugs or feature requests through the issue tracker
at [https://github.com/robn/HTTP-Tiny-Paranoid/issues](https://github.com/robn/HTTP-Tiny-Paranoid/issues).
You will be notified automatically of any progress on your issue.

## Source Code

This is open source software. The code repository is available for
public review and contribution under the terms of the license.

[https://github.com/robn/HTTP-Tiny-Paranoid](https://github.com/robn/HTTP-Tiny-Paranoid)

    git clone https://github.com/robn/HTTP-Tiny-Paranoid.git

# AUTHORS

- Robert Norris <rob@eatenbyagrue.org>

# COPYRIGHT AND LICENSE

This software is copyright (c) 2014 by Robert Norris.

This is free software; you can redistribute it and/or modify it under
the same terms as the Perl 5 programming language system itself.


Changes
{{$NEXT}}

0.07      2017-10-30 17:42:57+11:00 Australia/Melbourne
- Don't require peer address to match resolved address

0.06      Oct 30 2017
- Always check the peer IP address

0.05      Nov 01 2016
- Require latest HTTP::Tiny to get overridable peer IP
- Pass through resolved IP via overridable peer IP

0.04      Sep 29 2015
          Protect against DNS rebind attacks

0.03      Jun 04 2014
          Declare dependency on HTTP::Tiny
          Doc fixes

0.02      Jun 03 2014
          Fix build

0.01      Jun 03 2014
          First release


lib/HTTP/Tiny/Paranoid.pm
package HTTP::Tiny::Paranoid;

use strict;
use warnings;

# ABSTRACT: A safer HTTP::Tiny

use HTTP::Tiny 0.070;
use parent 'HTTP::Tiny';
use Net::DNS::Paranoid;
use Class::Method::Modifiers;

my $dns = Net::DNS::Paranoid->new;

sub blocked_hosts { shift; $dns->blocked_hosts(@_) }
sub whitelisted_hosts { shift; $dns->whitelisted_hosts(@_) }

around _open_handle => sub {
  my $next = shift;
  my $self = shift;

  my ($req, $scheme, $host, $port, $peer) = @_;

  if ($peer) {
    my ($ips, $error) = $dns->resolve($peer);
    die "$peer: $error\n" if defined $error;
    $self->$next($req, $scheme, $host, $port, $peer);
  }
  else {
    my ($ips, $error) = $dns->resolve($host);
    die "$host: $error\n" if defined $error;
    die "$host: no IP address found\n" unless @$ips;
    $self->$next($req, $scheme, $host, $port, $ips->[0]);
  }
};

1;
__END__

=pod

=encoding UTF-8

=head1 NAME

HTTP::Tiny::Paranoid - A safer HTTP::Tiny

=head1 SYNOPSIS

    # use just like HTTP::Tiny
    use HTTP::Tiny::Paranoid;
    my $response = HTTP::Tiny::Paranoid->new->get('http://example.com');

    # block or whitelist specific hosts
    # delegates to Net::DNS::Paranoid
    HTTP::Tiny::Paranoid->blocked_hosts([...]);
    HTTP::Tiny::Paranoid->whitelisted_hosts([...]);

=head1 DESCRIPTION

This module is a subclass of HTTP::Tiny that performs exactly one additional
function: before connecting, it passes the hostname to
L<Net::DNS::Paranoid>. If the hostname is rejected, then the request
is aborted before a connect is even attempted.

By default, L<Net::DNS::Paranoid> rejects connections to private network
ranges. The blocklist & whitelist can be manipulated using the C<blocked_hosts>
and C<whitelisted_hosts> class methods.

=head1 SEE ALSO

=over 4

=item *

L<HTTP::Tiny>

=item *

L<Net::DNS::Paranoid>

=item *

L<LWPx::ParanoidAgent>

=item *

L<LWP::UserAgent::Paranoid>

=back

=head1 SUPPORT

=head2 Bugs / Feature Requests

Please report any bugs or feature requests through the issue tracker
at L<https://github.com/robn/HTTP-Tiny-Paranoid/issues>.
You will be notified automatically of any progress on your issue.

=head2 Source Code

This is open source software. The code repository is available for
public review and contribution under the terms of the license.

L<https://github.com/robn/HTTP-Tiny-Paranoid>

  git clone https://github.com/robn/HTTP-Tiny-Paranoid.git

=head1 AUTHORS

=over 4

=item *

Robert Norris <rob@eatenbyagrue.org>

=back

=head1 COPYRIGHT AND LICENSE

This software is copyright (c) 2014 by Robert Norris.

This is free software; you can redistribute it and/or modify it under
the same terms as the Perl 5 programming language system itself.

=cut


Makefile.PL
# This file was automatically generated by Dist::Zilla::Plugin::MakeMaker v6.010.
use strict;
use warnings;



use ExtUtils::MakeMaker;

my %WriteMakefileArgs = (
  "ABSTRACT" => "A safer HTTP::Tiny",
  "AUTHOR" => "Robert Norris <rob\@eatenbyagrue.org>",
  "CONFIGURE_REQUIRES" => {
    "ExtUtils::MakeMaker" => 0
  },
  "DISTNAME" => "HTTP-Tiny-Paranoid",
  "LICENSE" => "perl",
  "NAME" => "HTTP::Tiny::Paranoid",
  "PREREQ_PM" => {
    "Class::Method::Modifiers" => 0,
    "HTTP::Tiny" => "0.070",
    "Net::DNS::Paranoid" => 0,
    "parent" => 0,
    "strict" => 0,
    "warnings" => 0
  },
  "VERSION" => "0.06",
  "test" => {
    "TESTS" => ""
  }
);


my %FallbackPrereqs = (
  "Class::Method::Modifiers" => 0,
  "HTTP::Tiny" => "0.070",
  "Net::DNS::Paranoid" => 0,
  "parent" => 0,
  "strict" => 0,
  "warnings" => 0
);


unless ( eval { ExtUtils::MakeMaker->VERSION(6.63_03) } ) {
  delete $WriteMakefileArgs{TEST_REQUIRES};
  delete $WriteMakefileArgs{BUILD_REQUIRES};
  $WriteMakefileArgs{PREREQ_PM} = \%FallbackPrereqs;
}

delete $WriteMakefileArgs{CONFIGURE_REQUIRES}
  unless eval { ExtUtils::MakeMaker->VERSION(6.52) };

WriteMakefile(%WriteMakefileArgs);
